### ISSUE TITLE
Document consensus rules from Zcash spec: 7.1.2 Transaction Consensus Rules

### ISSUE THREAD

**Issue by mpguerra** [ZcashFoundation/zebra#3223]
## TODO

Open a [zcash/zips](https://github.com/zcash/zips/) PR to split the spec section into multiple sub-sections. That way, there will be links to each sub-section, and the rules will be easier to find.

## Tasks

For each consensus rule:

- check that the rule is implemented in Zebra
- make sure the latest version of the rule is quoted in the docs for the code that implements it
- make sure the code implements the latest version of the rule

### Split 

- [x] #3425
- [x] #3426
- [x] #3427
- [x] #3428 

## Consensus rules

https://zips.z.cash/protocol/protocol.pdf#txnconsensus

![Screenshot 2021-12-16 at 16 04 14](https://user-images.githubusercontent.com/1311133/146396165-04810f4e-20f3-424d-a081-a04e0501ffb9.png)
![Screenshot 2021-12-16 at 16 05 32](https://user-images.githubusercontent.com/1311133/146396436-ccf27a53-6428-4a4e-8af4-1989601bf975.png)
![Screenshot 2021-12-16 at 16 05 49](https://user-images.githubusercontent.com/1311133/146396440-66b7210b-b036-4160-babe-d649d15151b7.png)

## Implementation
 

- #1917 
    - PRs: 
        - #3069
        - #3076
- #1980
    - PRs:
        - #2229
        - #2234
        - #2236
        - #2246  
- #1981
    - PRs:
        - #2371 
        - #2437  
- #2103
    - PR: #2441 
- #2105 
    - PR: #3039

**Comment by teor2345** [comment]
@mpguerra this is a very large task, so I'd like to split it between multiple people.

We could open a [zcash/zips](https://github.com/zcash/zips/) PR to split the spec section into multiple sub-sections?
That way there will be links to the sub-sections.

**Comment by mpguerra** [comment]
> @mpguerra this is a very large task, so I'd like to split it between multiple people.

We can possibly split it between the people that worked on the listed issues/PRs. Before we start working on this we should double check that there aren't any other issues/PRs that contributed towards implementing this section of the spec.

**Comment by mpguerra** [comment]
It would also be nice if these issues could be used by the auditors as a reference and so we should make sure the description is as complete and accurate as possible

**Comment by teor2345** [comment]
> > @mpguerra this is a very large task, so I'd like to split it between multiple people.
> 
> We can possibly split it between the people that worked on the listed issues/PRs. Before we start working on this we should double check that there aren't any other issues/PRs that contributed towards implementing this section of the spec.

Thanks, please re-assign me to some of the sub-tasks after you've split this task.

Here is the split I'd suggest, because it matches the structure of Zebra's code (and also the spec):
- transaction header (including expiry height)
- transaction size (including item counts)
- coinbase transactions (including coinbase-specific item counts and values)
- shielded pools: the rule is specific to a shielded pool (this is different from the network upgrade range of the rule, even if the network upgrade has the same name as a shielded pool)

> It would also be nice if these issues could be used by the auditors as a reference and so we should make sure the description is as complete and accurate as possible

This is unusual - I've never seen a project that keeps closed tickets up to date as the code changes. So I assumed that closed tickets are out of scope for the audits. Tickets are based on what we knew at the time they were closed. But we want to audit the current state of the software, so we need something that stays up to date.

So I think we should update the documentation next to the code we're getting audited. That way, we can maintain that documentation as we change the code, so it will be ready for future audits.

I don't know if this is related, but was it confusing to use open tickets to track questions for the auditors?

Let's talk about auditing closed tickets and tracking auditor questions at the Engineering sync.

**Comment by mpguerra** [comment]
> > It would also be nice if these issues could be used by the auditors as a reference and so we should make sure the description is as complete and accurate as possible
> 
> This is unusual - I've never seen a project that keeps closed tickets up to date as the code changes. So I assumed that closed tickets are out of scope for the audits. Tickets are based on what we knew at the time they were closed. But we want to audit the current state of the software, so we need something that stays up to date.
> 
> So I think we should update the documentation next to the code we're getting audited. That way, we can maintain that documentation as we change the code, so it will be ready for future audits.

I wasn't suggesting that we update the closed tickets. I think the tickets as they are can contain useful information about how and where a consensus rule was implemented in addition to the documentation within the code. This will obviously only be valid for the NU5 network upgrade and I assume we'll do an audit before any new network upgrade is scheduled.

**Comment by teor2345** [comment]
All the sub-tickets in this ticket have been completed.
